File: src/lib/clipboard.ts
import { invoke } from "@tauri-apps/api/core"
import { readText, writeText } from "@tauri-apps/plugin-clipboard-manager"

export async function clipboardWriteText(text: string): Promise<void> {
  try {
    await writeText(text)
    return
  } catch {
    /* fall through */
  }
  await invoke("clipboard_write_text", { text })
}

export async function clipboardReadText(): Promise<string> {
  try {
    const text = await readText()
    if (text) return text
  } catch {
    /* fall through */
  }
  try {
    return await invoke<string>("clipboard_read_text")
  } catch {
    return ""
  }
}

File: src-tauri/src/commands/clipboard.rs
//! System clipboard IPC (uses tauri-plugin-clipboard-manager on a background thread).

use tauri::AppHandle;
use tauri_plugin_clipboard_manager::ClipboardExt;

use crate::error::AppError;

#[tauri::command]
pub async fn clipboard_write_text(app: AppHandle, text: String) -> Result<(), AppError> {
    let app = app.clone();
    tauri::async_runtime::spawn_blocking(move || {
        app.clipboard()
            .write_text(text)
            .map_err(|e| AppError::Internal {
                message: e.to_string(),
            })
    })
    .await
    .map_err(|e| AppError::Internal {
        message: e.to_string(),
    })?
}

#[tauri::command]
pub async fn clipboard_read_text(app: AppHandle) -> Result<String, AppError> {
    let app = app.clone();
    tauri::async_runtime::spawn_blocking(move || {
        app.clipboard().read_text().map_err(|e| AppError::Internal {
            message: e.to_string(),
        })
    })
    .await
    .map_err(|e| AppError::Internal {
        message: e.to_string(),
    })?
}
